File: security/src/main/java/top/bettercode/summer/security/repository/RedisApiTokenRepository.java
package top.bettercode.summer.security.repository;

import java.lang.reflect.Method;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.redis.connection.RedisConnection;
import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.util.ClassUtils;
import org.springframework.util.ReflectionUtils;
import org.springframework.util.StringUtils;
import top.bettercode.summer.security.ApiToken;

public final class RedisApiTokenRepository implements ApiTokenRepository {

  private final Logger log = LoggerFactory.getLogger(RedisApiTokenRepository.class);
  private static final String API_AUTH = "api_auth:";
  private static final String ID = "id:";
  private static final String ACCESS_TOKEN = "access_token:";
  private static final String REFRESH_TOKEN = "refresh_token:";

  private final String keyPrefix;

  private static final boolean springDataRedis_2_0 = ClassUtils.isPresent(
      "org.springframework.data.redis.connection.RedisStandaloneConfiguration",
      RedisApiTokenRepository.class.getClassLoader());

  private final RedisConnectionFactory connectionFactory;
  private final JdkSerializationSerializer jdkSerializationSerializer = new JdkSerializationSerializer();

  private Method redisConnectionSet_2_0;

  public RedisApiTokenRepository(RedisConnectionFactory connectionFactory) {
    this(connectionFactory, "");
  }

  public RedisApiTokenRepository(RedisConnectionFactory connectionFactory, String prefix) {
    this.connectionFactory = connectionFactory;
    if (StringUtils.hasText(prefix)) {
      this.keyPrefix = API_AUTH + prefix + ":";
    } else {
      this.keyPrefix = API_AUTH;
    }
    if (springDataRedis_2_0) {
      this.loadRedisConnectionMethods_2_0();
    }
  }

  private void loadRedisConnectionMethods_2_0() {
    this.redisConnectionSet_2_0 = ReflectionUtils.findMethod(
        RedisConnection.class, "set", byte[].class, byte[].class);
  }

  private RedisConnection getConnection() {
    return connectionFactory.getConnection();
  }

  private byte[] serializeKey(String object) {
    return (this.keyPrefix + object).getBytes(StandardCharsets.UTF_8);
  }


  @Override
  public void save(ApiToken apiToken) {
    String scope = apiToken.getScope();
    String username = apiToken.getUsername();
    String id = scope + ":" + username;

    byte[] auth = jdkSerializationSerializer.serialize(apiToken);

    byte[] accessKey = serializeKey(ACCESS_TOKEN + apiToken.getAccessToken().getTokenValue());
    byte[] refreshKey = serializeKey(
        REFRESH_TOKEN + apiToken.getRefreshToken().getTokenValue());
    byte[] idKey = serializeKey(ID + id);

    try (RedisConnection conn = getConnection()) {
      ApiToken exist = getApiToken(idKey, conn);
      conn.openPipeline();
      //删除已存在
      if (exist != null) {
        byte[] existAccessKey = serializeKey(ACCESS_TOKEN + exist.getAccessToken().getTokenValue());
        byte[] existRefreshKey = serializeKey(
            REFRESH_TOKEN + exist.getRefreshToken().getTokenValue());
        if (!Arrays.equals(existAccessKey, accessKey)) {
          conn.keyCommands().del(existAccessKey);
        }
        if (!Arrays.equals(existRefreshKey, refreshKey)) {
          conn.keyCommands().del(existRefreshKey);
        }
      }

      if (springDataRedis_2_0) {
        try {
          this.redisConnectionSet_2_0.invoke(conn, accessKey, idKey);
          this.redisConnectionSet_2_0.invoke(conn, refreshKey, idKey);
          this.redisConnectionSet_2_0.invoke(conn, idKey, auth);
        } catch (Exception ex) {
          conn.closePipeline();
          throw new RuntimeException(ex);
        }
      } else {
        conn.stringCommands().set(accessKey, idKey);
        conn.stringCommands().set(refreshKey, idKey);
        conn.stringCommands().set(idKey, auth);
      }

      int access_expires_in = apiToken.getAccessToken().getExpires_in();
      int refresh_expires_in = apiToken.getRefreshToken().getExpires_in();
      conn.keyCommands().expire(accessKey, access_expires_in);
      conn.keyCommands().expire(refreshKey, refresh_expires_in);
      conn.keyCommands().expire(idKey, Math.max(access_expires_in, refresh_expires_in));

      conn.closePipeline();
    }
  }

  @Override
  public void remove(ApiToken apiToken) {
    String scope = apiToken.getScope();
    String username = apiToken.getUsername();
    String id = scope + ":" + username;
    byte[] accessKey = serializeKey(ACCESS_TOKEN + apiToken.getAccessToken().getTokenValue());
    byte[] refreshKey = serializeKey(
        REFRESH_TOKEN + apiToken.getRefreshToken().getTokenValue());
    byte[] idKey = serializeKey(ID + id);
    try (RedisConnection conn = getConnection()) {
      conn.openPipeline();
      conn.keyCommands().del(accessKey);
      conn.keyCommands().del(refreshKey);
      conn.keyCommands().del(idKey);
      conn.closePipeline();
    }
  }

  @Override
  public void remove(String scope, String username) {
    String id = scope + ":" + username;
    byte[] idKey = serializeKey(ID + id);
    try (RedisConnection conn = getConnection()) {
      ApiToken apiAuthenticationToken = getApiToken(idKey, conn);
      if (apiAuthenticationToken != null) {
        conn.openPipeline();
        byte[] accessKey = serializeKey(
            ACCESS_TOKEN + apiAuthenticationToken.getAccessToken().getTokenValue());
        byte[] refreshKey = serializeKey(
            REFRESH_TOKEN + apiAuthenticationToken.getRefreshToken().getTokenValue());
        conn.keyCommands().del(accessKey);
        conn.keyCommands().del(refreshKey);
        conn.keyCommands().del(idKey);
        conn.closePipeline();
      }
    }
  }

  @Override
  public ApiToken findByScopeAndUsername(String scope, String username) {
    String id = scope + ":" + username;
    byte[] idKey = serializeKey(ID + id);
    return findByIdKey(idKey);
  }

  @Nullable
  private ApiToken findByIdKey(byte[] idKey) {
    try (RedisConnection conn = getConnection()) {
      return getApiToken(idKey, conn);
    }
  }

  @Nullable
  private ApiToken getApiToken(byte[] idKey, RedisConnection conn) {
    byte[] bytes = conn.stringCommands().get(idKey);
    if (JdkSerializationSerializer.isEmpty(bytes)) {
      return null;
    }
    try {
      return (ApiToken) jdkSerializationSerializer.deserialize(bytes);
    } catch (Exception e) {
      log.warn("apiToken反序列化失败", e);
      try {
        conn.keyCommands().del(idKey);
      } catch (Exception ex) {
        log.warn("apiToken删除失败", ex);
      }
      return null;
    }
  }

  @Override
  public ApiToken findByAccessToken(String accessToken) {
    byte[] accessKey = serializeKey(ACCESS_TOKEN + accessToken);
    try (RedisConnection conn = getConnection()) {
      byte[] bytes = conn.stringCommands().get(accessKey);
      if (JdkSerializationSerializer.isEmpty(bytes)) {
        return null;
      }
      return getApiToken(bytes, conn);
    }
  }

  @Override
  public ApiToken findByRefreshToken(String refreshToken) {
    byte[] refreshKey = serializeKey(REFRESH_TOKEN + refreshToken);
    try (RedisConnection conn = getConnection()) {
      byte[] bytes = conn.stringCommands().get(refreshKey);
      if (JdkSerializationSerializer.isEmpty(bytes)) {
        return null;
      }
      return getApiToken(bytes, conn);
    }
  }
}

File: security/src/main/kotlin/top/bettercode/summer/security/repository/RedisApiTokenRepository.kt
package top.bettercode.summer.security.repository

import org.slf4j.LoggerFactory
import org.springframework.data.redis.connection.RedisConnection
import org.springframework.data.redis.connection.RedisConnectionFactory
import org.springframework.data.redis.connection.RedisPipelineException
import org.springframework.util.ClassUtils
import org.springframework.util.ReflectionUtils
import org.springframework.util.StringUtils
import top.bettercode.summer.security.token.ApiToken
import java.lang.reflect.Method
import java.nio.charset.StandardCharsets
import kotlin.math.max

class RedisApiTokenRepository @JvmOverloads constructor(private val connectionFactory: RedisConnectionFactory, prefix: String = "") : ApiTokenRepository {
    private val log = LoggerFactory.getLogger(RedisApiTokenRepository::class.java)
    private var keyPrefix: String? = null
    private val jdkSerializationSerializer = JdkSerializationSerializer()
    private var redisconnectionset20: Method? = null

    init {
        keyPrefix = if (StringUtils.hasText(prefix)) {
            "$API_AUTH$prefix:"
        } else {
            API_AUTH
        }
        if (springdataredis20) {
            loadredisconnectionmethods20()
        }
    }

    private fun loadredisconnectionmethods20() {
        redisconnectionset20 = ReflectionUtils.findMethod(
                RedisConnection::class.java, "set", ByteArray::class.java, ByteArray::class.java)
    }

    private val connection: RedisConnection
        get() = connectionFactory.connection

    private fun <T> RedisConnection.use(run: (RedisConnection) -> T?): T? {
        try {
            return run(this)
        } finally {
            this.close()
        }
    }

    private fun serializeKey(`object`: String): ByteArray {
        return (keyPrefix + `object`).toByteArray(StandardCharsets.UTF_8)
    }

    override fun save(apiToken: ApiToken) {
        try {
            val scope = apiToken.scope
            val username = apiToken.username
            val id = "$scope:$username"
            val auth = jdkSerializationSerializer.serialize(apiToken)
            val accessKey = serializeKey(ACCESS_TOKEN + apiToken.accessToken.tokenValue)
            val refreshKey = serializeKey(
                    REFRESH_TOKEN + apiToken.refreshToken.tokenValue)
            val idKey = serializeKey(ID + id)
            connection.use { conn ->
                val exist = getApiToken(idKey, conn)
                conn.openPipeline()
                //删除已存在
                if (exist != null) {
                    val existAccessKey = serializeKey(
                            ACCESS_TOKEN + exist.accessToken.tokenValue)
                    val existRefreshKey = serializeKey(
                            REFRESH_TOKEN + exist.refreshToken.tokenValue)
                    if (!existAccessKey.contentEquals(accessKey)) {
                        conn.keyCommands().del(existAccessKey)
                    }
                    if (!existRefreshKey.contentEquals(refreshKey)) {
                        conn.keyCommands().del(existRefreshKey)
                    }
                }
                if (springdataredis20) {
                    try {
                        redisconnectionset20!!.invoke(conn, accessKey, idKey)
                        redisconnectionset20!!.invoke(conn, refreshKey, idKey)
                        redisconnectionset20!!.invoke(conn, idKey, auth)
                    } catch (ex: Exception) {
                        conn.closePipeline()
                        throw RuntimeException(ex)
                    }
                } else {
                    conn.stringCommands()[accessKey] = idKey
                    conn.stringCommands()[refreshKey] = idKey
                    conn.stringCommands()[idKey] = auth
                }
                val accessExpiresIn = apiToken.accessToken.expires_in
                val refreshExpiresIn = apiToken.refreshToken.expires_in
                conn.keyCommands().expire(accessKey, accessExpiresIn.toLong())
                conn.keyCommands().expire(refreshKey, refreshExpiresIn.toLong())
                conn.keyCommands().expire(idKey, max(accessExpiresIn, refreshExpiresIn).toLong())
                conn.closePipeline()
            }
        } catch (e: Exception) {
            throw RuntimeException("保存授权信息失败", e)
        }
    }

    override fun remove(apiToken: ApiToken) {
        try {
            val scope = apiToken.scope
            val username = apiToken.username
            val id = "$scope:$username"
            val accessKey = serializeKey(ACCESS_TOKEN + apiToken.accessToken.tokenValue)
            val refreshKey = serializeKey(
                    REFRESH_TOKEN + apiToken.refreshToken.tokenValue)
            val idKey = serializeKey(ID + id)
            connection.use { conn ->
                conn.openPipeline()
                conn.keyCommands().del(accessKey)
                conn.keyCommands().del(refreshKey)
                conn.keyCommands().del(idKey)
                conn.closePipeline()
            }
        } catch (e: Exception) {
            throw RuntimeException("移除授权信息失败", e)
        }
    }

    override fun remove(scope: String, username: String) {
        try {
            val id = "$scope:$username"
            val idKey = serializeKey(ID + id)
            connection.use { conn ->
                val apiAuthenticationToken = getApiToken(idKey, conn)
                if (apiAuthenticationToken != null) {
                    conn.openPipeline()
                    val accessKey = serializeKey(
                            ACCESS_TOKEN + apiAuthenticationToken.accessToken.tokenValue)
                    val refreshKey = serializeKey(
                            REFRESH_TOKEN + apiAuthenticationToken.refreshToken.tokenValue)
                    conn.keyCommands().del(accessKey)
                    conn.keyCommands().del(refreshKey)
                    conn.keyCommands().del(idKey)
                    conn.closePipeline()
                }
            }
        } catch (e: RedisPipelineException) {
            throw RuntimeException("移除授权信息失败", e)
        }
    }

    override fun remove(scope: String, usernames: List<String>) {
        usernames.forEach { remove(scope, it) }
    }

    override fun findByScopeAndUsername(scope: String, username: String): ApiToken? {
        val id = "$scope:$username"
        val idKey = serializeKey(ID + id)
        return findByIdKey(idKey)
    }

    private fun findByIdKey(idKey: ByteArray): ApiToken? {
        return connection.use { conn -> return@use getApiToken(idKey, conn) }
    }

    private fun getApiToken(idKey: ByteArray, conn: RedisConnection): ApiToken? {
        return try {
            val bytes = conn.stringCommands()[idKey]
            if (JdkSerializationSerializer.isEmpty(bytes)) {
                return null
            }
            try {
                jdkSerializationSerializer.deserialize(bytes!!) as ApiToken
            } catch (e: Exception) {
                log.warn("apiToken反序列化失败", e)
                try {
                    conn.keyCommands().del(idKey)
                } catch (ex: Exception) {
                    log.warn("apiToken删除失败", ex)
                }
                null
            }
        } catch (e: Exception) {
            throw RuntimeException("查询授权信息失败", e)
        }
    }

    override fun findByAccessToken(accessToken: String): ApiToken? {
        try {
            val accessKey = serializeKey(ACCESS_TOKEN + accessToken)
            return connection.use { conn ->
                val bytes = conn.stringCommands()[accessKey]
                return@use if (JdkSerializationSerializer.isEmpty(bytes)) {
                    null
                } else getApiToken(bytes!!, conn)
            }
        } catch (e: Exception) {
            throw RuntimeException("查询授权信息失败", e)
        }
    }

    override fun findByRefreshToken(refreshToken: String): ApiToken? {
        try {
            val refreshKey = serializeKey(REFRESH_TOKEN + refreshToken)
            return connection.use { conn ->
                val bytes = conn.stringCommands()[refreshKey]
                return@use if (JdkSerializationSerializer.isEmpty(bytes)) {
                    null
                } else getApiToken(bytes!!, conn)
            }
        } catch (e: Exception) {
            throw RuntimeException("查询授权信息失败", e)
        }
    }

    companion object {
        private const val API_AUTH = "api_auth:"
        private const val ID = "id:"
        private const val ACCESS_TOKEN = "access_token:"
        private const val REFRESH_TOKEN = "refresh_token:"
        private val springdataredis20 = ClassUtils.isPresent(
                "org.springframework.data.redis.connection.RedisStandaloneConfiguration",
                RedisApiTokenRepository::class.java.classLoader)
    }
}
